FreeRTOS/Test/Target/tests/smp/disable_preemption/disable_preemption.c

```diff
@@ -0,0 +1,241 @@
+
+/**
+ * @brief Test case "Disable Preemption".
+ */
+void Test_DisablePreemption( void );
+
+/**
+ * @brief Disable preemption test task.
+ */
+static void prvTestPreemptionDisableTask( void * pvParameters );
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Handles of the tasks created in this test.
+ */
+static TaskHandle_t xTaskHandles[ configNUMBER_OF_CORES + 1 ];
+
+/**
+ * @brief Indexes of the tasks created in this test.
+ */
+static uint32_t xTaskIndexes[ configNUMBER_OF_CORES + 1 ];
+
+/**
+ * @brief Flags to indicate the test result.
+ */
+static BaseType_t xTestResult = pdFAIL;
+/*-----------------------------------------------------------*/
+
+static void prvTestPreemptionDisableTask( void * pvParameters )
+{
+    uint32_t currentTaskIdx = *( ( uint32_t * ) pvParameters );
+    uint32_t taskIndex;
+    eTaskState taskState;
+    BaseType_t xAllHighPriorityTasksSuspended = pdFALSE;
+
+    if( currentTaskIdx < configNUMBER_OF_CORES )
+    {
+        /* Tasks with smaller index have higher priority. Higher priority tasks
+         * suspend themselves and are resumed later by the lowest priority task
+         * after the lower priority task disables preemption for itself. */
+        vTaskSuspend( NULL );
+    }
+    else
+    {
+        /* Wait for all the other higher priority tasks to suspend themselves. */
+        while( xAllHighPriorityTasksSuspended == pdFALSE )
+        {
+            for( taskIndex = 0; taskIndex < configNUMBER_OF_CORES; taskIndex++ )
+            {
+                taskState = eTaskGetState( xTaskHandles[ taskIndex ] );
+
+                if( taskState != eSuspended )
+                {
+                    break;
+                }
+            }
+
+            if( taskIndex == configNUMBER_OF_CORES )
+            {
+                xAllHighPriorityTasksSuspended = pdTRUE;
+            }
+        }
+
+        /* Disable preemption and resume all the other higher priority tasks.
+         * At this point, the number of higher priority ready tasks is equal
+         * to the number of cores. Still this lower priority must not be
+         * switched out because it has disabled preemption for itself. */
+        vTaskPreemptionDisable( NULL );
+
+        for( taskIndex = 0; taskIndex < configNUMBER_OF_CORES; taskIndex++ )
+        {
+            vTaskResume( xTaskHandles[ taskIndex ] );
+        }
+
+        /* This task must not be switched out for any other higher priority
+         * ready task because it has disabled preemption for itself. The
+         * execution of the next line ensures that this task is not switched out
+         * even though a higher priority ready task is available. This variable
+         * is checked in the test runner. */
+        xTestResult = pdPASS;
+    }
+
+    /* Busy looping here to occupy this core. */
+    for( ; ; )
+    {
+        /* Always running, put asm here to avoid optimization by compiler. */
+        __asm volatile ( "nop" );
```

Comment:
I see the following patter in one other test:
```
#ifndef portNOP
    #define TEST_NOP()    __asm volatile ( "nop" )
#else
    #define TEST_NOP    portNOP
#endif
```

Should we do the same here too for consistency?

Reply:
Use TEST_NOP macro for testing.